### ISSUE TITLE
Usage and installation

### ISSUE THREAD

**Issue by verhoop** [bkbilly/medisanabp_ble#1]
Installed it via HACS, rebooted the installation to complete.
How do I fail search? (Any debug/output log?)
I'm not seeing any new devices coming up.

**Comment by verhoop** [comment]
When adding the integration there are a lot of Bluetooth devices to choose from. (Around 80) possible all the detected BLE devices. How do I know which one is the Blood pressure monitor?

**Comment by bkbilly** [comment]
You can enable debugging from the devices page under the **Medisana Blood Pressure BLE** integration there is a button to **Enable debug logging**.

The device is detected automatically, so you don't have to choose from this list.
In order to be discovered on Home Assistant, the blood pressure device should show the Bluetooth icon on the screen and it should be near your Bluetooth Proxy.
I am using the **BU 530 connect** and the icon is shown after I take a measurement or if I long press the power button. This should be similar to other models, but I haven't used any other to test it out.

**Comment by verhoop** [comment]
Hi, Im using a BU 580 connect, but seems like its the same firmware. (also BT symbol after measurement and hold down to resend)
What kind of a BT proxy are you using? Im using a Shelly 1 plus

**Comment by bkbilly** [comment]
I believe that it should work on all Medisana Blood Pressure measurements the same way.
You can use any Bluetooth proxy as long as it supports active connections.
I've tried it on the internal Bluetooth of Raspberry and an ESP32 flashed with ESPHome.

**Comment by verhoop** [comment]
Ok, when you do the measurement, do you have the vitadock app open as well? Normally Bluetooth needs to pair.Perhaps the ShellyPlus is not supporting active connections.What did you do on the raspberry? //PeterOn 9 Apr 2024, at 16:38, Vasilis Koulis ***@***.***> wrote:﻿
I believe that it should work on all Medisana Blood Pressure measurements the same way.
You can use any Bluetooth proxy as long as it supports active connections.
I've tried it on the internal Bluetooth of Raspberry and an ESP32 flashed with ESPHome.

**Comment by bkbilly** [comment]
You shouldn't have the app open because only one device can be connected. The pairing process is done automatically by the integration I created.
I am using a Raspberry Pi 4B that has Bluetooth on board and I've installed Home Assistant Operating System.
To enable the Bluetooth I added the [Bluetooth](https://www.home-assistant.io/integrations/bluetooth/) integration under the devices page.

**Comment by verhoop** [comment]
Wondering… have you tried connection it with the Raspberry Pi’s Bluetooth disabled as well? (So only ESP32)? //PeterOn 9 Apr 2024, at 20:48, Vasilis Koulis ***@***.***> wrote:﻿
You shouldn't have the app open because only one device can be connected. The pairing process is done automatically by the integration I created.
I am using a Raspberry Pi 4B that has Bluetooth on board and I've installed Home Assistant Operating System.
To enable the Bluetooth I added the Bluetooth integration under the devices page.

**Comment by bkbilly** [comment]
Yeah, I've tried both ways and it works.

**Comment by verhoop** [comment]
Thanks, it was the shelly plus device which didnt want to play along. 
I ordered an ESP32 and flashed it with Bluetooth Proxy and it worked flawlessly. :)
PS, I got a Medisana Scale as well, there are some integrations done but via python and MQTT.
Would be nice to incorporate it as well.  I will try to see if I can alter your code and apply it to the scale as well. (im more an infrastructure guy, not developer)

PPS, Strange.. The initial reading went well. (values popped up in the sensor), second reading the BLM gave me an error E4 (meaning no bluetooth communication. (also with the push start when sleeping part)

**Comment by bkbilly** [comment]
Medisana devices depend on subscription to Bluetooth messages which don't work as well as the rest of the BLE integrations of Home Assistant. I've also encountered error codes, but usually the readings are recognized. I don't know how fix this to make sure the values are better recognized.

I found [this](https://github.com/keptenkurk/BS440/blob/master/BS440.py) that might help you find on which topics you should subscribe. Unfortunately I don't have a Medisana scale to help you out. I am using a Xiaomi scale and it works flawlessly.

**Comment by verhoop** [comment]
Thanks, Got another one tho :

ValueError: Invalid datetime: sensor.18_7a_93_e9_1e_46_1e46_measured_date has timestamp device class but provides state 2024-04-10T21:53:00+02:00:<class 'str'> resulting in ''str' object has no attribute 'tzinfo''
2024-04-11 08:08:27.817 ERROR (MainThread) [homeassistant.components.sensor] Error while setting up medisanabp_ble platform for sensor


alueError: Invalid datetime: sensor.18_7a_93_e9_1e_46_1e46_measured_date has timestamp device class but provides state 2024-04-10T21:53:00+02:00:<class 'str'> resulting in ''str' object has no attribute 'tzinfo''
2024-04-11 08:08:27.817 ERROR (MainThread) [homeassistant.components.sensor] Error while setting up medisanabp_ble platform for sensor
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/sensor/__init__.py", line 600, in state
    if value.tzinfo is None:
       ^^^^^^^^^^^^
AttributeError: 'str' object has no attribute 'tzinfo'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 356, in _async_setup_platform
    await asyncio.shield(awaitable)
  File "/config/custom_components/medisanabp_ble/sensor.py", line 119, in async_setup_entry
    coordinator.async_register_processor(processor, SensorEntityDescription)
  File "/usr/src/homeassistant/homeassistant/components/bluetooth/passive_update_processor.py", line 341, in async_register_processor
    processor.async_register_coordinator(self, entity_description_class)
  File "/usr/src/homeassistant/homeassistant/components/bluetooth/passive_update_processor.py", line 475, in async_register_coordinator
    self.async_update_listeners(data)
  File "/usr/src/homeassistant/homeassistant/components/bluetooth/passive_update_processor.py", line 567, in async_update_listeners
    update_callback(data)
  File "/usr/src/homeassistant/homeassistant/components/bluetooth/passive_update_processor.py", line 677, in _handle_processor_update
    self.async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 998, in async_write_ha_state
    self._async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1119, in _async_write_ha_state
    state, attr, capabilities, shadowed_attr = self.__async_calculate_state()
                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1056, in __async_calculate_state
    state = self._stringify_state(available)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1004, in _stringify_state
    if (state := self.state) is None:
                 ^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/sensor/__init__.py", line 611, in state
    raise ValueError(
ValueError: Invalid datetime: sensor.18_7a_93_e9_1e_46_1e46_measured_date has timestamp device class but provides state 2024-04-10T21:53:00+02:00:<class 'str'> resulting in ''str' object has no attribute 'tzinfo''
2024-04-11 08:08:27.897 WARNING (MainThread) [homeassistant.components.daikin] Connection failed for 192.168.50.41
2024-04-11 08:13:04.803 ERROR (MainThread) [homeassistant] Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/bleak_retry_connector/__init__.py", line 505, in _async_wrap_bluetooth_connection_error_retry
    return await func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^

**Comment by verhoop** [comment]
Same as this? 
https://github.com/home-assistant/core/pull/73502/commits/7b7de62aefb9a473b5a197c7b0377ea3db0c2615

**Comment by bkbilly** [comment]
This is already done on my parser [here](https://github.com/bkbilly/medisanabp_ble/blob/main/custom_components/medisanabp_ble/medisana_bp/parser.py#L78). 
I can't think of any reason of why you get this error.

**Comment by smartgate-one** [comment]
My Medisana BU 570 will not be automaticly found. Maybe the Manufacterer_id Advertising is named 1872B instead of 18498 ?

If I add this manually only one actual Dataset will be stored.

**Comment by bkbilly** [comment]
You can try changing it on the file `manifest.json` and restart Home Assistant. 
If it works, I could add support for the Manufacturer ID of your blood preasure measurement device.

**Comment by ghost** [comment]
Hi, I have also a BU 570 and have the same issue as @smartgate-one.

Changing the Manufacterer_id did not seem to make a difference. At least in regards receiving records. not sure on the auto-detect.

Here are some logs:

![Log1](https://github.com/bkbilly/medisanabp_ble/assets/172963881/32620d70-3f3c-47c1-881b-5f263fde6bf9)
![Log2](https://github.com/bkbilly/medisanabp_ble/assets/172963881/f153ccfe-bba5-4b3c-a305-18efb9dd1578)

I then changed the parser to:
`
        try:

            local_timezone = datetime.now(timezone.utc).astimezone().tzinfo
            datetime_str = f"{dyear}/{dmonth}/{dday} {dhour}:{dminu:0>2} {local_timezone}"
            date = datetime.strptime(datetime_str, '%Y/%m/%d %H:%M %Z')
            self.update_sensor(
                key=str(MedisanaBPSensor.TIMESTAMP),
                native_unit_of_measurement=None,
                native_value=date.replace(tzinfo=local_timezone),
                name="Measured Date",
            )
        except:
            _LOGGER.error("Can't add Measured Date")
`
Which at least removed the error message around timezone. But the second error is still there:

![Log3](https://github.com/bkbilly/medisanabp_ble/assets/172963881/453ab4a2-81ff-4919-b076-b35721d32866)

I assume the the device is sending also historic values, the first received (old) record is parsed and then something goes wrong.

**Comment by LukevdPalen** [comment]
See my pull request: https://github.com/bkbilly/medisanabp_ble/pull/2

BU 570 is detected automatically, but HA has difficulty connecting to the device.

**Comment by 7542** [comment]
do not understand how to fix:
Please please help

```
Logger: custom_components.medisanabp_ble
Quelle: components/bluetooth/active_update_processor.py:125
Integration: Medisana Blood Pressure BLE ([Dokumentation](https://github.com/bkbilly/medisanabp_ble), [Probleme](https://github.com/bkbilly/medisanabp_ble/issues))
Erstmals aufgetreten: 11:14:37 (2 Vorkommnisse)
Zuletzt protokolliert: 11:19:35

<same MAC-Address>: Bluetooth error whilst polling: <same MAC-Address> - <same MAC-Address>: Failed to connect after 8 attempt(s): Timeout waiting for connect response while connecting to <same MAC-Address> after 20.0s, disconnect timed out: False, after 20.0s
<same MAC-Address>: Bluetooth error whilst polling: <same MAC-Address> - <same MAC-Address>: Failed to connect after 4 attempt(s): Timeout waiting for connect response while connecting to <same MAC-Address> after 20.0s, disconnect timed out: False, after 20.0s
```

**Comment by duceduc** [comment]
@verhoop 
> Hi, Im using a BU 580 connect, but seems like its the same firmware. (also BT symbol after measurement and hold down to resend)
What kind of a BT proxy are you using? Im using a Shelly 1 plus

I am assuming you have got it working and this integration does support the BU 580???

@bkbilly 
> I am using a Xiaomi scale and it works flawlessly.

What model is your Xiaomi scale and you are able to use the same integration to capture the reading of the scale????

**Comment by bkbilly** [comment]
My xiaomi scale is not working with this integration. I am using the xiaomi integration

**Comment by aornt** [comment]
There is also a problem with my BU570. It only displays the 9th value in the history of blood pressure measurements.
When I change the python code a little as following in the parser.py to get the complete content of the 'data' variable:

```
    @retry_bluetooth_connection_error()
    def notification_handler(self, _, data) -> None:
        """Helper for command events"""
        syst = data[2] * 256 + data[1]
        diast = data[4] * 256 + data[3]
        arter = data[6] * 256 + data[5]
        dyear = data[8] * 256 + data[7]
        dmonth = data[9]
        dday = data[10]
        dhour = data[11]
        dminu = data[12]
        x13   = data[13]
        puls = data[15] * 256 + data[14]
        user = data[16]
        x17 = data[17]
        x18 = data[18]
        datalen = len(data)
        
        try:
            datetime_str = f"{dyear}/{dmonth}/{dday} {dhour}:{dminu:0>2}"
            date = datetime.strptime(datetime_str, '%Y/%m/%d %H:%M')
            local_timezone = datetime.now(timezone.utc).astimezone().tzinfo
            self.update_sensor(
                key=str(MedisanaBPSensor.TIMESTAMP),
                native_unit_of_measurement=None,
                native_value=date.replace(tzinfo=local_timezone),
                name="Measured Date",
            )
        except:
            _LOGGER.error("Can't add Measured Date")

        _LOGGER.info(
            "Got data from BPM device (syst: %s, diast: %s, puls: %s)",
            syst, diast, puls)

        _LOGGER.info(
            "date data from BPM device (dyear: %s, dmonth: %s, dday: %s)",
            dyear, dmonth, dday)
            
        _LOGGER.info(
            "time data from BPM device (dhour: %s, dminu: %s, x13: %s)",
            dhour, dminu, x13)

        _LOGGER.info(
            "more data from BPM device (arter: %s, user: %s, x17: %s, x18: %s, datalen: %d)",
            arter, user, x17, x18, datalen)
            
        self.update_sensor(
            key=str(MedisanaBPSensor.SYSTOLIC),
            native_unit_of_measurement=Units.PRESSURE_MMHG,
            native_value=syst,
            device_class=SensorDeviceClass.PRESSURE,
            name="Systolic",
        )
        self.update_sensor(
            key=str(MedisanaBPSensor.DIASTOLIC),
            native_unit_of_measurement=Units.PRESSURE_MMHG,
            native_value=diast,
            device_class=SensorDeviceClass.PRESSURE,
            name="Diastolic",
        )
        self.update_sensor(
            key=str(MedisanaBPSensor.PULSE),
            native_unit_of_measurement="bpm",
            native_value=puls,
            name="Pulse",
        )
        self._event.set()
        return

```
I get the following result when I turn logging on 

 

> 2024-11-04 15:07:11.268 DEBUG (MainThread) [custom_components.medisanabp_ble.medisana_bp.parser] Parsing  MedisanaBP BLE advertisement data: <BluetoothServiceInfoBleak name=1872B address=A4:C1:38:AC:4C:E2 rssi=-91 manufacturer_data={} service_data={} service_uuids=['00001810-0000-1000-8000-00805f9b34fb'] source=24:4C:AB:00:8C:8C connectable=True time=1003459.645585388 tx_power=None>
> 2024-11-04 15:07:11.269 DEBUG (MainThread) [custom_components.medisanabp_ble.medisana_bp.parser] Connecting to BLE device: A4:C1:38:AC:4C:E2
> 2024-11-04 15:07:17.251 INFO (MainThread) [custom_components.medisanabp_ble.medisana_bp.parser] Got data from BPM device (syst: 173, diast: 106, puls: 66)
> 2024-11-04 15:07:17.252 INFO (MainThread) [custom_components.medisanabp_ble.medisana_bp.parser] date data from BPM device (dyear: 2024, dmonth: 11, dday: 4)
> 2024-11-04 15:07:17.252 INFO (MainThread) [custom_components.medisanabp_ble.medisana_bp.parser] time data from BPM device (dhour: 2, dminu: 41, x13: 0)
> 2024-11-04 15:07:17.252 INFO (MainThread) [custom_components.medisanabp_ble.medisana_bp.parser] more data from BPM device (arter: 139, user: 1, x17: 0, x18: 0, datalen: 19)
> 2024-11-04 15:07:17.253 ERROR (MainThread) [homeassistant] Error doing job: Task exception was never retrieved (None)
> Traceback (most recent call last):
>   File "/usr/local/lib/python3.12/site-packages/bleak_retry_connector/__init__.py", line 505, in _async_wrap_bluetooth_connection_error_retry
>     return await func(*args, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^
> TypeError: object NoneType can't be used in 'await' expression
> 2024-11-04 15:07:17.316 DEBUG (MainThread) [custom_components.medisanabp_ble.medisana_bp.parser] Disconnected from active bluetooth client



I'm a bit stuck on how to get the first 8 measurements. It seems the 'data' variable only has 1 set of measurements. There are much more measurements, about 50 or so. It's a bit odd that only number 9 shows up
If I take another measurement the values shift and I read the former 8th value that has become number 9 now

Maybe something flows over at number 10

My bluetooth proxy is an Olimex.  (olimex-esp32-poe-iso) that is installed with  ESPHome 2024.10.2

Does any one have a clue, or maybe an idea how to get rid of the error message ?

**Comment by aornt** [comment]
sorry for the bold letterype , I can't get it off

**Comment by 7542** [comment]
i have Problems getting data from my "old" BU530. Only once data receiving by many messurements!
Ordered new BU570 ... but same issue. Please can someone explain me where to get logging data of BLE in HASSIO?

**Comment by aornt** [comment]
I have both the BU570 and the BU530 but I can't get neither of them working.
 Also I tried with a bluetooth sniffer on the phone. There is quite some info available in the communication with the medisana but because there is only a short time with the connection it is very troublesome to investigate what is available. 

And then still there is the point that I don't know how the datatransfer between the olimex and the medisana app is working. 
So i'm also interested in the answer to your question.

**Comment by aornt** [comment]
If your question was  how to get the logs in HomeAssistant:
Settings > Devices & services > integrations > medisana bloodpressure BLE > enable debug logging

to view the logs: 

Settings > system > Logs > show raw logs

I also have logging in ESPhome for the olimex:

Esphome-dashboard > Bluetooth Proxy 008c8c> logs

But my esphome is on a seperate Proxmox VM, you may have an other kind of connection to your medisana bpm

**Comment by bethgocs** [comment]
I have a BT580 and had to change the manufacture_Id to 31256. 

I have also the problem, that the measurements are only read once with full history. If I do some measurements and later transfer the whole history only the last value is transfered.

Additional HA does transfer my measured values until I restarted the integration.

**Comment by bkbilly** [comment]
It is not possible to fill historical data of Home Assistant sensors, so this integration reads the latest value.

**Comment by bethgocs** [comment]
Yes, you're right. Sorry for my ignorance. I have found a custom integration that promises to work: https://github.com/patrickvorgers/Home-Assistant-Import-Energy-Data

May this will be helpful.

**Comment by bkbilly** [comment]
Unfortunately this can't be done with a home assistant integration.
The one you suggested stops any data to the DB and adds data manually. It only supports specific databases.
This can't be implemented on my integration, so the only option is to have only the latest value.

**Comment by bethgocs** [comment]
Thanks anyway :)
